tests/cocotb/basic/test/test_hwpe_stream_merge.py

```diff
@@ -0,0 +1,305 @@
+            push_ready[j] = dut.ready_i[j].value
+            cocotb.log.info(f'push_i[{j}].ready.value = {push_ready[j]}')
+
+        #-----------------------------------
+        # Logging data
+        #-----------------------------------
+        for j in range(NB_IN_STREAMS-1,-1,-1):
+
+            # Stimuli setting for NB_IN_STREAMS inputs
+            cocotb.log.info(f'===== INPUT LOG =====')
+
+            # Record log
+            cocotb.log.info(f'----- INPUT STREAM # {j} -----')
+            cocotb.log.info(f'push_i[{j}].data.value  = {hex(push_data[j])}')
+            cocotb.log.info(f'push_i[{j}].valid.value = {   push_valid[j] }')
+            cocotb.log.info(f'push_i[{j}].strb.value  = {bin(push_strb[j])}')
+            
+        
+        cocotb.log.info(f'----- OUTPUT READY -----')
+        cocotb.log.info(f'pop_o.ready.value     = {pop_ready}')
+        
+        # Extract outputs
+        cocotb.log.info(f'===== OUTPUT LOG =====')
+
+        cocotb.log.info(f'pop_o.valid.value     = {pop_valid}')
+        cocotb.log.info(f'pop_o.data.value      = {hex(pop_data)}')
+        cocotb.log.info(f'pop_o.strb.value      = {bin(pop_strb)}')
+
+        #-----------------------------------
+        # Assertion checks
+        #-----------------------------------
+
+        # Checking for output mismatch
+        # Expected output should be a simple concatenation of inputs
+        # Note that MSB is higher number in array
+        data_check  = 0
+        strb_check  = 0
+        valid_check = 0
+        ready_flag  = 1
+
+        for j in range(NB_IN_STREAMS-1,-1,-1):
+
+            data_check   = (data_check << DATA_WIDTH) + push_data[j]
+            strb_check   = (strb_check << STRB_WIDTH) + push_strb[j]
+            valid_check += push_valid[j]
+
+            if(pop_ready != push_ready[j]):
+                ready_flag = 0
+
+        # Checking for data mismatch
+        assert(data_check == pop_data), f"ERROR! Output mismatch - Expected output: {hex(data_check)}; Actual output: {hex(pop_data)}"
+
+        # Checking for strb mismatch
+        assert(strb_check == pop_strb), f"ERROR! Output mismatch - Expected output: {bin(strb_check)}; Actual output: {bin(pop_strb)}"
+
+        # Checking for valid mismatch
+        valid_expect = math.floor(valid_check/NB_IN_STREAMS)
+        assert(valid_expect == pop_valid), f"ERROR! Valid mismatch - Expected valid: {valid_expect}; Actual output: {pop_valid}"
+
+        # Checking for ready mismatch
+        assert (ready_flag == 1), f"ERROR! Ready mismatch - Double check ready values. Output ready input should be the same for input ready signals."
+
+
+#-----------------------------------
+# Pytest run
+#-----------------------------------
+
+# Parametrization
+@pytest.mark.parametrize(
+"parameters", [
+    {"DATA_WIDTH": str(DATA_WIDTH), "NB_IN_STREAMS": str(NB_IN_STREAMS)}
+])
+
+# Main test run
+def test_hwpe_stream_merge(parameters):
+
+    global rtl_sources
```

Comment:
Why are you defining these variables up top if you don't use them earlier? I would just define them locally here instead. Also using globals is a bit of a bad practice. Do you really need to define these variables globally?

Reply:
I wanted to put them up top so that the modifiable parameters are collated and it's the first thing users can see. I will take out the global, I thought Python does not see variables declared outside. Apparently they do just like what I did for some parameters.